Update 0292-EEPROM-use.md (#5)
PlanktoScope/proposals: design/00292: EEPROM storage.md

# Proposal: Use the PlanktoScope HAT EEPROM chip to store hardware configuration

> Author(s): Vallet, Laurent; Le Corre, Cléa Last updated: 30/10/2024 Discussion at https://github.com/PlanktoScope/PlanktoScope/issues/292

Reviewer: ```suggestion
Author(s): Vallet, Laurent; Le Corre, Cléa

Last updated: 2024-10-30

Discussion at https://github.com/PlanktoScope/PlanktoScope/issues/292
```

## Abstract

> The following document proposes to store the PlanktoScope (PS) hardware configuration in the 32 Kbits EEPROM chip of the PlanktoScope HAT in order to : 

Reviewer: ```suggestion
The following document proposes to store PlanktoScope hardware configuration metadata in the 32 Kbit EEPROM chip of the PlanktoScope HAT, in order to: 
```
(Thibaut has asked that we avoid using "ps" or "pkscope" as an abbreviation for "PlanktoScope")

> * automatically save the hardware state of the machine

Reviewer: ```suggestion
* pre-populate the machine with hardware-specific metadata as part of manufacturing at FairScope
* persist hardware-specific metadata across re-flashes of the Raspberry Pi's SD card image
```

* easily retrieve the hardware configuration for debugging

> * feed the metadata with precise hardware information

Reviewer: ```suggestion
* provide correct hardware information as part of the metadata which gets exported with datasets generated by PlanktoScopes
```

## Background

A specific HAT was developped to satisfy the PlanktoScope needs. 
This PlanktoScope HAT connected to the raspberry pi and all of the external devices (led, motors, pump, camera) contains a 32 Kbit EEPROM chip which is a "type of non-volatile ROM that enables individual bytes of data to be erased and reprogrammed" (Wikipedia).
It is therefore possible to store some data inside this chip and to read it back.

> For now (Sunray version 2024.0.0-beta.2) some of the hardware configuration of the PlanktoScope is entered by the user in the GUI.

Reviewer: ```suggestion
Currently (as of PlanktoScope OS v2023.9.0 and v2024.0.0), the hardware configuration of the PlanktoScope is entered by the user in the Node-RED dashboard and is stored in a `hardware.json` file which is lost when the PlanktoScope's SD card is re-flashed with a new SD card image.
```

We would like to have an hardware configuration specific to each machine in case some specific modification is made.
We would like to store this configuration in a safe place, independent from the software installed on the raspberry pi.

> ## Proposal

Reviewer: The proposal should also describe how data will be laid out on the EEPROM. For example, will individual metadata fields be serialized/deserialized and stored at separate addresses? Or will you instead have a giant JSON string containing all metadata fields, which you would read/write all together from the EEPROM. And how will you represent values in the EEPROM's bytes, i.e. what will be your serialization format, especially for multi-byte values?

Ideally, you could briefly summarize some advantages and disadvantages of the different possible approaches.

> We therefore propose to write a python script that allows the storage and the reading of the PlanktoScope hardware information on the EEPROM chip.

Reviewer: ```suggestion
We therefore propose to add functionality to PlanktoScope OS to store and load the PlanktoScope's hardware metadata onto/from the EEPROM chip.
```

This information stored should contain : 
* PlanktoScope reference
* Plantoscope serial number
* PlanktoScope version
* Plantoscope fabrication date
* HAT serial number
* HAT version
* Driver reference
* Pump reference
* Stepper focus reference
* Objective lens reference
* Tube lens reference
* Flowcell thickness
* LED reference

Those informations could be used to feed the metadata stored in the hardware.json file as described in the issue 290 (https://github.com/PlanktoScope/PlanktoScope/issues/290).

> 

Reviewer: Since you are adding MQTT API routes, please also describe the design of the MQTT API routes you are adding.

## Rationale

The EEPROM chip currently used for the HAT has a capacity of 32Kbits. This provides adequate space for the necessary configuration without redundancy or excessive overhead.
The use of our HAT is relevant as the PlanktoScope HAT is linked to every device of the machine. This allows unique data storage for each PlanktoScope unit, helping track individual hardware changes or updates.
The EEPROM on the HAT is accessible directly through the I2C bus of the Raspberry Pi, simplifying software development.

This solution is exclusive to the PlanktoScope HAT, limiting compatibility with other configurations (e.g., Adafruit HATs without EEPROM).
Though currently sufficient, 32Kbits could become a constraint if future features or expanded data storage needs arise.

## Compatibility

> The main compatibility issue concerns the adafruit HAT users.

Reviewer: ```suggestion
The main backwards-compatibility issue concerns users of older Adafruit HAT-based versions of the PlanktoScope design.
```

> For those, a simple solution would be to continue selecting their configuration in the GUI, or adapt the script to store informations in the Raspberry pi EEPROM chip in case of not finding the PlanktoScope HAT EEPROM adress.

Reviewer: Based on https://www.raspberrypi.com/documentation/computers/raspberry-pi.html , I believe the Raspberry Pi's EEPROM chip may already be used/reserved by the Raspberry Pi as part of how it boots up. If this is a correct understanding of the Raspberry Pi, then we cannot use the Raspberry Pi's EEPROM chip to store our own data.

When there is no PlanktoScope HAT with EEPROM, the other alternative would be to use a JSON file on the SD card as a substitute for storing data on EEPROM. I recommend you to describe this option in this design document.

The PlanktoScope versions V2.5 and V2.6 do not natively support writing data to the EEPROM chip because the write-protect (WP) pin of the EEPROM is enabled. To disable write protection, a solder point must be applied to the WP pin.

> 

Reviewer: This design document also needs to talk about data schema compatibility: if/when the metadata we store on the EEPROM changes (e.g. by adding or removing fields, or by adding new enum values for strings), how will we ensure backwards compatibility and/or forwards compatibility? I believe a common approach is to store a data schema version number as part of the data, as a way to detect when we might need to handle certain backwards/forwards-compatibility situations in certain ways.

Since there will be an MQTT API associated with this functionality, I also recommend addressing the question of API compatibility. For example, maybe this design document say the following things in the compatibility section:
1. This proposal will not modify any existing MQTT API routes, but it will add some new MQTT API routes related to EEPROM functionality.
2. All MQTT API routes added by this proposal are to be considered experimental, with no guarantees of compatibility in the future.

## Implementation

> The EEPROM project has reached a functional stage, with all desired actions for writing and reading hardware information now operational. The next phase will focus on integrating this functionality into the broader PlanktoScope software, allowing users to access and manage their device's hardware information throughout its lifecycle. This integration will enhance the user experience, offering consistent hardware management for improved maintenance and lifecycle tracking of each PlanktoScope device.

Reviewer: ```suggestion
The EEPROM project has reached a functional prototype, with all desired actions for writing and reading hardware information now operational.
The next phase will focus on integrating this functionality into the broader PlanktoScope software, allowing users to access and manage their device's hardware information throughout its lifecycle.
This integration will enhance the user experience, offering consistent hardware management for improved maintenance and lifecycle tracking of each PlanktoScope device.
```

## Open issues (if applicable)
One current limitation in this project is the requirement to manually solder a connection on the Write Protect (WP) pin of the EEPROM chip to enable writing operations. This step introduces some complexity and potential error for users or technicians during the assembly or modification process.

> To address this, a planned improvement is to incorporate a bridge between the WP pin and GPIO pin 4 on the Raspberry Pi. By controlling the write protection via software, this enhancement would streamline write operations by allowing the program to enable or disable the WP functionality dynamically. This improvement would enhance flexibility and reduce the need for physical alterations.

Reviewer: ```suggestion
To address this, a planned improvement is to incorporate a bridge between the WP pin and GPIO pin 4 on the Raspberry Pi. By controlling the write protection via software, this enhancement would streamline write operations by allowing the program to enable or disable the WP functionality dynamically. This improvement would enhance flexibility and reduce the need for physical alterations.

Currently, no design or plan has been proposed for enabling use of a JSON file on the PlanktoScope's SD card as a substitute for EEPROM (see the "Compatibility" section's discussion of backwards-compatibilitly with Adafruit HAT-based PlanktoScopes). That question will be left for future work, maybe as part of this proposal.
```
